File: day8/part2.ts
import * as fs from "fs";

type Point = { x: number; y: number; z: number };
type Edge = { dist2: number; a: number; b: number };

class UnionFind {
  parent: number[];
  rank: number[];
  components: number;

  constructor(n: number) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.rank = new Array(n).fill(0);
    this.components = n;
  }

  find(x: number): number {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }
    return this.parent[x];
  }

  union(a: number, b: number): boolean {
    const ra = this.find(a);
    const rb = this.find(b);
    if (ra === rb) return false;

    if (this.rank[ra] < this.rank[rb]) this.parent[ra] = rb;
    else if (this.rank[ra] > this.rank[rb]) this.parent[rb] = ra;
    else {
      this.parent[rb] = ra;
      this.rank[ra]++;
    }

    this.components--;
    return true;
  }
}

const input = fs.readFileSync("input.txt", "utf-8").trim().split("\n");

const points: Point[] = input.map((line) => {
  const [x, y, z] = line.split(",").map(Number);
  return { x, y, z };
});

const edges: Edge[] = [];

for (let i = 0; i < points.length; i++) {
  for (let j = i + 1; j < points.length; j++) {
    const dx = points[i].x - points[j].x;
    const dy = points[i].y - points[j].y;
    const dz = points[i].z - points[j].z;
    edges.push({ dist2: dx * dx + dy * dy + dz * dz, a: i, b: j });
  }
}

edges.sort((a, b) => a.dist2 - b.dist2);

const uf = new UnionFind(points.length);

for (const e of edges) {
  if (uf.union(e.a, e.b) && uf.components === 1) {
    console.log(points[e.a].x * points[e.b].x);
    break;
  }
}

File: day8/part2.py
from __future__ import annotations
from dataclasses import dataclass
from typing import List, Tuple


@dataclass(frozen=True)
class Point:
    x: int
    y: int
    z: int


class UnionFind:
    def __init__(self, size: int) -> None:
        self.parent = list(range(size))
        self.rank = [0] * size
        self.components = size

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False

        if self.rank[ra] < self.rank[rb]:
            self.parent[ra] = rb
        elif self.rank[ra] > self.rank[rb]:
            self.parent[rb] = ra
        else:
            self.parent[rb] = ra
            self.rank[ra] += 1

        self.components -= 1
        return True


def read_input(path: str) -> List[Point]:
    points: List[Point] = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            x, y, z = map(int, line.strip().split(","))
            points.append(Point(x, y, z))
    return points


def solve(points: List[Point]) -> int:
    edges: List[Tuple[int, int, int]] = []

    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            dx = points[i].x - points[j].x
            dy = points[i].y - points[j].y
            dz = points[i].z - points[j].z
            dist2 = dx * dx + dy * dy + dz * dz
            edges.append((dist2, i, j))

    edges.sort(key=lambda e: e[0])

    uf = UnionFind(len(points))

    for _, a, b in edges:
        if uf.union(a, b) and uf.components == 1:
            return points[a].x * points[b].x

    raise RuntimeError("Graph never fully connected")


if __name__ == "__main__":
    points = read_input("input.txt")
    print(solve(points))
